src/neuroconv/tools/roiextractors/roiextractors.py

```diff
@@ -586,6 +600,170 @@ def image_mask_iterator():
+) -> NWBFile:
+    """
+    Adds the fluorescence traces specified by the metadata to the nwb file.
+    The fluorescence traces that are added are the one located in metadata["Ophys"]["Fluorescence"].
+    The df/F traces that are added are the one located in metadata["Ophys"]["DfOverF"].
+
+    Parameters
+    ----------
+    segmentation_extractor : SegmentationExtractor
+        The segmentation extractor to get the traces from.
+    nwbfile : NWBFile
+        The nwbfile to add the fluorescence traces to.
+    metadata : dict
+        The metadata for the fluorescence traces.
+    plane_index : int, optional
+        The index of the plane to add the fluorescence traces to.
+
+    Returns
+    -------
+    NWBFile
+        The nwbfile passed as an input with the fluorescence traces added.
+    """
+
+    # Set the defaults and required infrastructure
+    metadata_copy = deepcopy(metadata)
+    default_metadata = get_default_ophys_metadata()
+    metadata_copy = dict_deep_update(default_metadata, metadata_copy, append_list=False)
+
+    # df/F metadata
+    dff_metadata = metadata_copy["Ophys"]["DfOverF"]
+    dff_name = dff_metadata["name"]
+
+    # Fluorescence traces metadata
+    fluorescence_metadata = metadata_copy["Ophys"]["Fluorescence"]
+    fluorescence_name = fluorescence_metadata["name"]
+
+    # Get traces from the segmentation extractor
+    traces_to_add = segmentation_extractor.get_traces_dict()
+
+    # Filter empty data
+    traces_to_add = {trace_name: trace for trace_name, trace in traces_to_add.items() if trace is not None}
+    # Filter all zero data
+    traces_to_add = {
+        trace_name: trace for trace_name, trace in traces_to_add.items() if any(x != 0 for x in np.ravel(trace))
+    }
+
+    # Early return if there is nothing to add
+    if not traces_to_add:
+        return nwbfile
+
+    # Create a reference for ROIs from the plane segmentation
+    roi_table_region = _create_roi_table_region(
+        segmentation_extractor=segmentation_extractor,
+        nwbfile=nwbfile,
+        metadata=metadata_copy,
+        plane_index=plane_index,
+    )
+
+    rate = (
+        np.nan
+        if segmentation_extractor.get_sampling_frequency() is None
+        else float(segmentation_extractor.get_sampling_frequency())
+    )
+
+    roi_response_series_kwargs = dict(
+        rois=roi_table_region,
+        # TODO: timestamps or rate (for timestamps we need frame_to_time method)
+        # TODO: check for regularity, only use timestamps if irregular timing
+        starting_time=0.0,
+        rate=rate,
+        unit="n.a.",
+    )
+
+    container = defaultdict(
+        lambda: _get_segmentation_data_interface(
+            nwbfile=nwbfile,
+            data_interface_name=fluorescence_name,
+        ),
+        # we expect that the df/F trace name is "Dff"
+        Dff=_get_segmentation_data_interface(
+            nwbfile=nwbfile,
+            data_interface_name=dff_name,
+        ),
+    )
+
```

Comment:
Right now, the function call which is the first argument of the defaultdict is called every time you use the dict. If you call the `_get_segmentation_data_interface` and then assign the value to the dict you can avoid this repeated call.

Reply:
the reason why I tried this is because given this example:

```
if "dff" in traces_to_add:
    dff = _get_segmentation_data_interface("DfOverF")
if remaining_traces:
    fluorescence = _get_segmentation_data_interface("Fluorescence")
x = defaultdict(lambda: fluorescence, Dff=dff) 
```
this will raise an error that local variable `dff` is referenced before assignment in those cases where there are no df over f trace to add. if you have an idea how to overcome this, I'm happy to change it!

Reply:
Something similar to updating typical dicts works. The only difference is that defaultdicts have this attribut `default_factor` that you can also set post initialization. 

A -probably non working- example of how this could go:
```
trace_to_container = defaultdict()
if remaining_traces:
   trace_to_container.default_factory = lambda: fluorescence # or lambda: call_to_your_function
if "dff" in traces_to_add:
    dff = _get_segmentation_data_interface("DfOverF")
    trace_to_container.update(Dff=dff)
```

That said, if you think that your version is more readable maybe you are right? I was just commenting on the small inefficiency of calling the function every time in the loop.